rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSource.java

```diff
@@ -284,13 +300,18 @@ public void snapshotState(FunctionSnapshotContext context) throws Exception {
 
         unionOffsetStates.clear();
 
-        if (LOG.isDebugEnabled()) {
-            LOG.debug("Snapshotted state, last processed offsets: {}, checkpoint id: {}, timestamp: {}",
-                offsetTable, context.getCheckpointId(), context.getCheckpointTimestamp());
-        }
+        HashMap<MessageQueue, Long> currentOffsets = new HashMap<>(offsetTable.size());
 
         for (Map.Entry<MessageQueue, Long> entry : offsetTable.entrySet()) {
             unionOffsetStates.add(Tuple2.of(entry.getKey(), entry.getValue()));
+            currentOffsets.put(entry.getKey(), entry.getValue());
+        }
+
+        pendingOffsetsToCommit.put(context.getCheckpointId(), currentOffsets);
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Snapshotted state, last processed offsets: {}, checkpoint id: {}, timestamp: {}",
+                offsetTable, context.getCheckpointId(), context.getCheckpointTimestamp());
```

Comment:
should truncate the map of pending offsets to commit, to prevent infinite growth,you can set fixed threshold value ```public static final int MAX_NUM_PENDING_CHECKPOINTS = 100;```